futures-sink/src/channel_impls.rs

```diff
@@ -1,44 +1,67 @@
-use {Async, Sink, AsyncSink, StartSend, Poll};
+use {Async, Sink, Poll};
 use futures_core::task;
 use futures_channel::mpsc::{Sender, SendError, UnboundedSender};
 
-fn res_to_async_sink<T>(res: Result<(), T>) -> AsyncSink<T> {
-    match res {
-        Ok(()) => AsyncSink::Ready,
-        Err(x) => AsyncSink::Pending(x),
+/// The error type of `<Sender<T> as Sink>`
+///
+/// It will contain a value of type `T` if one was passed to `start_send`
+/// after the channel was closed.
+#[derive(Debug)]
+pub struct ChannelClosed<T>(Option<T>);
```

Comment:
How come this was needed?

Reply:
Previously, it was always the case that an error contained a `T`. That's not the case anymore, since `poll_ready` doesn't have a `T` to respond with (previously, the only function that could error was `start_send`, which has a `T`).

Reply:
Hm could we mirror the `Option` in the original futures-channel crate as well though? (to avoid the introduction of another type)

Reply:
Sure-- i debated about this, since it's sort of misleading to the end user, because some functions will only ever return `Some` or `None`. In the `Sink` trait that's not preventable, but it is if you're directly using functions on `Channel`. I think I agree that the extra complexity isn't worth it, though.

Reply:
Eh I'd probably be fine either way, I understand now what's going on at least :)

Reply:
I guess the key question to me is whether we can expect `start_send` to *always* be able to give you the value back on an error, in which case the `Sink` trait should probably continue to use a wrapper around the error type.

cc @carllerche